python-frontend/src/main/java/org/sonar/python/types/v2/LazyUnionType.java

```diff
@@ -16,15 +16,18 @@
  */
 package org.sonar.python.types.v2;
 
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Set;
+import java.util.stream.Stream;
 
 public class LazyUnionType implements PythonType, ResolvableType {
 
-  private final Set<PythonType> candidates = new HashSet<>();
+  private final Set<PythonType> candidates;
 
   public LazyUnionType(Set<PythonType> candidates) {
-    this.candidates.addAll(candidates);
+    this.candidates = candidates.stream().flatMap(LazyUnionType::flattenLazyUnionTypes).collect(HashSet::new, HashSet::add, HashSet::addAll);
```

Comment:
there is `Collectors.toCollection(HashSet::new)` which would do the same thing (see https://stackoverflow.com/a/30082600)